File: RateLimiter2/FixedWindowCounter.js
const RateLimiterStrategy = require('./RateLimiterStrategy');

class FixedWindowCounter extends RateLimiterStrategy {
  limit = 0;
  windowMs = 0;
  requests = new Map();
  windowStart = Date.now();

  constructor(limit, windowMs) {
    super();
    this.limit = limit;
    this.windowMs = windowMs;
    this.windowStart = Date.now();
  }

  isAllowed(ip) {
    let currentTime = Date.now();
    if (currentTime - this.windowStart >= this.windowMs) {
      this.windowStart = currentTime;
      this.requests = new Map();
    }

    if (!this.requests.has(ip)) {
      this.requests.set(ip, 1);
      return true;
    }

    let currentCount = this.requests.get(ip);
    if (currentCount >= this.limit) {
      return false;
    }

    this.requests.set(ip, currentCount + 1);
    return true;
  }
}

module.exports = FixedWindowCounter;

File: Go/RateLimiter2-go/fixed_window_counter.go
package main

import (
	"sync"
	"time"
)

type FixedWindowCounter struct {
	mu          sync.Mutex
	Limit       int
	WindowMs    int64
	Requests    map[string]int
	WindowStart int64
}

func NewFixedWindowCounter(limit int, windowMs int64) *FixedWindowCounter {
	return &FixedWindowCounter{
		Limit:       limit,
		WindowMs:    windowMs,
		Requests:    make(map[string]int),
		WindowStart: time.Now().UnixMilli(),
	}
}

func (f *FixedWindowCounter) IsAllowed(ip string) bool {
	f.mu.Lock()
	defer f.mu.Unlock()
	currentTime := time.Now().UnixMilli()
	if currentTime-f.WindowStart >= f.WindowMs {
		f.WindowStart = currentTime
		f.Requests = make(map[string]int)
	}
	if _, ok := f.Requests[ip]; !ok {
		f.Requests[ip] = 1
		return true
	}
	if f.Requests[ip] >= f.Limit {
		return false
	}
	f.Requests[ip]++
	return true
}
